src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,399 @@
+            var inOrder = true; // assume the arguments are in order in default
+
+            for (var i = 0; i < arguments.Count; i++)
+            {
+                // Parameter index cannot out of its range, #arguments is larger than #parameter only if the last parameter with keyword params
+                var parameterIndex = Math.Min(i, parameters.Length - 1);
+
+                // If the argument has a name, get the corresponding parameter index
+                var nameSyntax = arguments[i].NameColon?.Name;
+                if (nameSyntax != null)
+                {
+                    var name = nameSyntax.ToString();
+                    var found = false;
+                    for (var j = 0; j < parameters.Length; j++)
+                    {
+                        if (name.Equals(parameters[j].Name))
+                        {
+                            // Check if the argument is in order with parameters.
+                            // If the argument breaks the order, the rest arguments of matched functions must have names
+                            if (i != j)
+                                inOrder = false;
+                            parameterIndex = j;
+                            found = true;
+                            break;
+                        }
+                    }
+                    if (!found) return false;
+                }
+
+                // The argument is either in order with parameters, or have a matched name with parameters
+                var argType = semanticModel.GetTypeInfo(arguments[i].Expression, cancellationToken);
+                if (argType.Type != null && (inOrder || nameSyntax is object))
+                {
+                    // The type of argument must be convertible to the type of parameter
+                    if (!parameters[parameterIndex].IsParams
+                        && semanticModel.Compilation.ClassifyCommonConversion(argType.Type, parameters[parameterIndex].Type).Exists)
+                    {
+                        if (matchedTypes[parameterIndex]) return false;
+                        matchedTypes[parameterIndex] = true;
+                    }
+                    else if (parameters[parameterIndex].IsParams
+                        && semanticModel.Compilation.ClassifyCommonConversion(argType.Type, parameters[parameterIndex].Type).Exists)
+                    {
+                        // The parameter with keyword params takes an array type, then it cannot be matched more than once
+                        if (matchedTypes[parameterIndex]) return false;
+                        matchedTypes[parameterIndex] = true;
+                        paramsMatchedByArray = true;
+                    }
+                    else if (parameters[parameterIndex].IsParams
+                             && parameters.Last().Type is IArrayTypeSymbol paramsType
```

Comment:
do we know that `parameters` is non-empty (and thus won't throw on `.Last()`)?

Reply:
right, `parameters` could be empty, I didn't think about this case